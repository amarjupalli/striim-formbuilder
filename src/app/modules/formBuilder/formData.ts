export const FORM_INPUTS = {
  TEXT: "text",
  CHECKBOX: "checkbox",
  SELECT: "select",
  MULTI_SELECT: "multiSelect",
} as const;
interface BaseFormField {
  name: string;
  label: string;
  required?: boolean;
}

interface TextFormField extends BaseFormField {
  type: typeof FORM_INPUTS.TEXT;
}

interface OptionsFormField<T extends string> extends BaseFormField {
  type:
    | typeof FORM_INPUTS.CHECKBOX
    | typeof FORM_INPUTS.MULTI_SELECT
    | typeof FORM_INPUTS.SELECT;
  options: T[];
}

export type FormField<T extends string> = TextFormField | OptionsFormField<T>;

const fields = [
  {
    type: FORM_INPUTS.TEXT,
    name: "firstName",
    label: "First Name",
    required: true,
  },
  {
    type: FORM_INPUTS.TEXT,
    name: "lastName",
    label: "Last Name",
    required: true,
  },
  {
    type: FORM_INPUTS.CHECKBOX,
    name: "isSoftwareEngineer",
    label: "Are you a Software Engineer?",
    options: ["yes", "no"],
  },
  {
    type: FORM_INPUTS.MULTI_SELECT,
    name: "proficiency",
    label: "Which languages are you proficient in?",
    options: ["Javascript", "Java", "Python", "CSS", "SQL", "None of them"],
  },
  {
    type: FORM_INPUTS.MULTI_SELECT,
    name: "officeDays",
    label: "When can you come to our office in Palo Alto?",
    options: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Never"],
    required: true,
  },
  {
    type: FORM_INPUTS.SELECT,
    name: "operatingSystem",
    label: "What's your favorite operating system to code on?",
    options: ["Mac", "Windows", "Linux"],
  },
] as const satisfies FormField<string>[];

export interface FormBuilderData<T extends string> {
  fields: FormField<T>[];
}

const formData = { fields } as const satisfies FormBuilderData<string>;

export default formData;
