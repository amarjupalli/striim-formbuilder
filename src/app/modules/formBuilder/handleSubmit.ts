function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
  e.preventDefault();
  console.log(
    "On Save, formValues:",
    new FormData(e.target as HTMLFormElement),
  );
}

export default handleSubmit;
