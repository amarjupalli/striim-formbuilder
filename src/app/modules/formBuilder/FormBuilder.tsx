"use client";
import React from "react";
import handleSubmit from "./handleSubmit";
import { type FormField, type FormData } from "./formData";

interface FormBuilderProps {
  formData: FormData<string>;
}

function renderFromField(field: FormField<string>) {
  if (field.type === "text") {
    return (
      <div className="field" key={`$form-field-${field.name}`}>
        <label htmlFor={field.name}>
          {field.label}
          <input
            type="text"
            name={field.name}
            id={field.name}
            placeholder={field.label}
          />
        </label>
      </div>
    );
  }

  if (field.type === "checkbox") {
    return (
      <div className="field" key={`$form-field-${field.name}`}>
        <div>{field.label}</div>
        <div style={{ display: "inline-flex", gap: "16px" }}>
          {field.options?.map((option) => (
            <div key={`input-checkbox-${option}`}>
              <label htmlFor={option}>{option}</label>
              <input
                type="checkbox"
                name={`${field.name}-${option}`}
                id={`${field.name}-${option}`}
              />{" "}
            </div>
          ))}
        </div>
      </div>
    );
  }

  if (field.type === "multiSelect" || field.type === "select") {
    const { label, name, options, type } = field;
    return (
      <div className="field" key={`$form-field-${name}`}>
        <label htmlFor={name}>{label}</label>
        <select name={name} id={name} multiple={type === "multiSelect"}>
          {options?.map((option) => (
            <option key={`${name}-option-${option}`} value={option}>
              {option}
            </option>
          ))}
        </select>
      </div>
    );
  }
}

export default function FormBuilder({ formData }: FormBuilderProps) {
  const { fields } = formData;
  return (
    <form className="form" name="applicantForm" onSubmit={handleSubmit}>
      {fields.map(renderFromField)}
      <button type="submit">Submit</button>
    </form>
  );
}
