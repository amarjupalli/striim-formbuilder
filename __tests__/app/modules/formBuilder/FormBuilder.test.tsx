import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import FormBuilder from "@/app/modules/formBuilder/FormBuilder";
import formData from "@/app/modules/formBuilder/formData";
import handleSubmit from "@/app/modules/formBuilder/actions";

jest.mock("../../../../src/app/modules/formBuilder/handleSubmit");

describe("FormBuilder", () => {
  it("renders the FormBuilder", () => {
    const { container } = render(<FormBuilder formData={formData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render text inputs and labels", () => {
    render(<FormBuilder formData={formData} />);
    expect(screen.getByText(/first name/i)).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: /first name/i }),
    ).toBeInTheDocument();
    expect(screen.getByText(/last name/i)).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: /last name/i }),
    ).toBeInTheDocument();
  });

  it("should render a submit and reset button", () => {
    render(
      <FormBuilder
        formData={{
          fields: [{ type: "text", name: "testInput", label: "Test input" }],
        }}
      />,
    );
    const submitButton = screen.getByRole("button", { name: /submit/i });
    const resetButton = screen.getByRole("button", { name: /reset/i });
    expect(submitButton).toBeInTheDocument();
    expect(resetButton).toBeInTheDocument();

    fireEvent.click(submitButton);
    expect(handleSubmit).toHaveBeenCalledTimes(1);
  });

  describe("radio buttons and checkboxes", () => {
    it("should render radio buttons and labels", () => {
      const { container } = render(<FormBuilder formData={formData} />);
      expect(
        screen.getByText("Are you a Software Engineer?"),
      ).toBeInTheDocument();

      expect(screen.getByText("yes")).toBeInTheDocument();
      expect(
        container.querySelector("#isSoftwareEngineer-yes"),
      ).toBeInTheDocument();

      expect(screen.getByText("no")).toBeInTheDocument();
      expect(
        container.querySelector("#isSoftwareEngineer-no"),
      ).toBeInTheDocument();
    });

    it("should render check boxes and labels", () => {
      const { container } = render(<FormBuilder formData={formData} />);
      expect(
        screen.getByText(/select your area of interests/i),
      ).toBeInTheDocument();

      expect(screen.getByText("Backend")).toBeInTheDocument();
      expect(
        container.querySelector("#interestedIn-Backend"),
      ).toBeInTheDocument();

      expect(screen.getByText("Frontend")).toBeInTheDocument();
      expect(
        container.querySelector("#interestedIn-Frontend"),
      ).toBeInTheDocument();
    });
  });

  it("should render select boxes, listboxes and their title", () => {
    render(<FormBuilder formData={formData} />);

    // assert for multiSelect aka ListBox
    expect(
      screen.getByText("Which languages are you proficient in?"),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("listbox", {
        name: "Which languages are you proficient in?",
      }),
    ).toBeInTheDocument();

    // assert for another multiSelect aka ListBox
    expect(
      screen.getByText("When can you come to our office in Palo Alto?"),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("listbox", {
        name: "When can you come to our office in Palo Alto?",
      }),
    ).toBeInTheDocument();

    // assert for regular select dropdown
    expect(
      screen.getByText("What's your favorite operating system to code on?"),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("combobox", {
        name: "What's your favorite operating system to code on?",
      }),
    ).toBeInTheDocument();
  });
});
