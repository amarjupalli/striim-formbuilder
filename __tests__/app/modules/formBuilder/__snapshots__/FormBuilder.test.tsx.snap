// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBuilder renders the FormBuilder 1`] = `
<div>
  <form
    class="form"
    name="applicantForm"
  >
    <div
      class="field"
    >
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="First Name"
        required=""
        type="text"
      />
    </div>
    <div
      class="field"
    >
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        required=""
        type="text"
      />
    </div>
    <div
      class="field"
    >
      <label>
        Are you a Software Engineer?
      </label>
      <div
        style="display: inline-flex; gap: 16px; flex-wrap: wrap;"
      >
        <div>
          <label
            for="yes"
          >
            yes
          </label>
          <input
            id="isSoftwareEngineer-yes"
            name="isSoftwareEngineer"
            type="radio"
            value="yes"
          />
           
        </div>
        <div>
          <label
            for="no"
          >
            no
          </label>
          <input
            id="isSoftwareEngineer-no"
            name="isSoftwareEngineer"
            type="radio"
            value="no"
          />
           
        </div>
      </div>
    </div>
    <div
      class="field"
    >
      <label>
        Select your area of interests
      </label>
      <div
        style="display: inline-flex; gap: 16px; flex-wrap: wrap;"
      >
        <div>
          <label
            for="Backend"
          >
            Backend
          </label>
          <input
            id="interestedIn-Backend"
            name="interestedIn"
            type="checkbox"
            value="Backend"
          />
           
        </div>
        <div>
          <label
            for="Frontend"
          >
            Frontend
          </label>
          <input
            id="interestedIn-Frontend"
            name="interestedIn"
            type="checkbox"
            value="Frontend"
          />
           
        </div>
        <div>
          <label
            for="Full stack"
          >
            Full stack
          </label>
          <input
            id="interestedIn-Full stack"
            name="interestedIn"
            type="checkbox"
            value="Full stack"
          />
           
        </div>
        <div>
          <label
            for="DevOps/Infra"
          >
            DevOps/Infra
          </label>
          <input
            id="interestedIn-DevOps/Infra"
            name="interestedIn"
            type="checkbox"
            value="DevOps/Infra"
          />
           
        </div>
      </div>
    </div>
    <div
      class="field"
    >
      <label
        for="proficiency"
      >
        Which languages are you proficient in?
      </label>
      <select
        id="proficiency"
        multiple=""
        name="proficiency"
      >
        <option
          value="Javascript"
        >
          Javascript
        </option>
        <option
          value="Java"
        >
          Java
        </option>
        <option
          value="Python"
        >
          Python
        </option>
        <option
          value="CSS"
        >
          CSS
        </option>
        <option
          value="SQL"
        >
          SQL
        </option>
        <option
          value="None of them"
        >
          None of them
        </option>
      </select>
    </div>
    <div
      class="field"
    >
      <label
        for="officeDays"
      >
        When can you come to our office in Palo Alto?
      </label>
      <select
        id="officeDays"
        multiple=""
        name="officeDays"
        required=""
      >
        <option
          value="Monday"
        >
          Monday
        </option>
        <option
          value="Tuesday"
        >
          Tuesday
        </option>
        <option
          value="Wednesday"
        >
          Wednesday
        </option>
        <option
          value="Thursday"
        >
          Thursday
        </option>
        <option
          value="Friday"
        >
          Friday
        </option>
        <option
          value="Never"
        >
          Never
        </option>
      </select>
    </div>
    <div
      class="field"
    >
      <label
        for="operatingSystem"
      >
        What's your favorite operating system to code on?
      </label>
      <select
        id="operatingSystem"
        name="operatingSystem"
      >
        <option
          value="Mac"
        >
          Mac
        </option>
        <option
          value="Windows"
        >
          Windows
        </option>
        <option
          value="Linux"
        >
          Linux
        </option>
      </select>
    </div>
    <div
      class="buttonContainer"
    >
      <button
        type="submit"
      >
        Submit
      </button>
      <button
        type="reset"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;
